File: problems/0024-swap-nodes-in-pairs/solution.js
/**
 * @problemNumber: 24
 * @problemName: Swap Nodes in Pairs
 * @difficulty: Medium
 * @topics: Linked List, Recursion
 * @companies: Microsoft, Amazon, Facebook
 */

function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

/**
 * Approach: Iterative
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 *
 * @param {ListNode} head
 * @return {ListNode}
 */
function swapPairs(head) {
    const dummy = new ListNode(0);
    dummy.next = head;
    let prev = dummy;

    while (prev.next && prev.next.next) {
        let first = prev.next;
        let second = prev.next.next;

        prev.next = second;
        first.next = second.next;
        second.next = first;

        prev = first;
    }

    return dummy.next;
}

// Helper to create list
function createList(arr) {
    let dummy = new ListNode(0);
    let curr = dummy;
    for (let val of arr) {
        curr.next = new ListNode(val);
        curr = curr.next;
    }
    return dummy.next;
}

// Helper to array
function toArray(head) {
    let res = [];
    while (head) {
        res.push(head.val);
        head = head.next;
    }
    return res;
}

// Test cases
let head = createList([1, 2, 3, 4]);
console.log(toArray(swapPairs(head))); // [2,1,4,3]

module.exports = { swapPairs };

File: problems/0024-swap-nodes-in-pairs/solution.py
"""
Problem Number: 24
Problem Name: Swap Nodes in Pairs
Difficulty: Medium
Topics: Linked List, Recursion
Companies: Microsoft, Amazon, Facebook
"""

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        Approach: Iterative
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        
        while prev.next and prev.next.next:
            first = prev.next
            second = prev.next.next
            
            # Swap
            prev.next = second
            first.next = second.next
            second.next = first
            
            # Move
            prev = first
            
        return dummy.next

# Helper methods for testing
def create_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for val in arr:
        curr.next = ListNode(val)
        curr = curr.next
    return dummy.next

def list_to_arr(head):
    res = []
    while head:
        res.append(head.val)
        head = head.next
    return res

# Test cases
if __name__ == "__main__":
    solution = Solution()
    head = create_list([1,2,3,4])
    res = solution.swapPairs(head)
    print(list_to_arr(res)) # [2,1,4,3]
